src/tile/worker.cc

```diff
@@ -0,0 +1,331 @@
+      // Skip degenerate lines (less than 2 points)
+      if (clipped_line.size() < 2) {
+        continue;
+      }
+
+      // Convert to vtzero points, removing consecutive duplicates
+      std::vector<vtzero::point> tile_coords;
+      tile_coords.reserve(clipped_line.size());
+
+      int32_t last_x = INT32_MIN;
+      int32_t last_y = INT32_MIN;
+
+      for (const auto& pt : clipped_line) {
+        int32_t x = static_cast<int32_t>(std::round(pt.x()));
+        int32_t y = static_cast<int32_t>(std::round(pt.y()));
+
+        // Skip consecutive duplicate points (can happen after rounding)
+        if (x == last_x && y == last_y) {
+          continue;
+        }
+
+        tile_coords.emplace_back(x, y);
+        last_x = x;
+        last_y = y;
+      }
+
+      // Must have at least 2 unique points to create a valid linestring
+      if (tile_coords.size() < 2) {
+        continue;
+      }
+
+      // Create linestring feature for this clipped segment
+      // Use GraphId as feature ID for global uniqueness and debugging
+      vtzero::linestring_feature_builder feature{layer_roads};
+      feature.set_id(static_cast<uint64_t>(edge_id));
+      feature.add_linestring_from_container(tile_coords);
+
+      // Add properties
+      // Get hierarchy level (shared for both directions)
+      uint32_t level = edge_id.level();
+      feature.add_property(key_level, vtzero::encoded_property_value(level));
+
+      // Determine direction suffix based on edge->forward()
+      std::string dir_suffix = edge->forward() ? "forward" : "reverse";
+
+      // Add this edge's properties with appropriate suffix
+      std::string edge_id_str = std::to_string(static_cast<uint64_t>(edge_id));
+      auto key_edge_id = edge->forward() ? key_edge_id_fwd : key_edge_id_rev;
+      auto key_road_class = edge->forward() ? key_road_class_fwd : key_road_class_rev;
+      auto key_use = edge->forward() ? key_use_fwd : key_use_rev;
+      auto key_tunnel = edge->forward() ? key_tunnel_fwd : key_tunnel_rev;
+      auto key_bridge = edge->forward() ? key_bridge_fwd : key_bridge_rev;
+      auto key_roundabout = edge->forward() ? key_roundabout_fwd : key_roundabout_rev;
+
+      feature.add_property(key_edge_id, vtzero::encoded_property_value(edge_id_str));
+      feature.add_property(key_road_class, vtzero::encoded_property_value(
+                                               static_cast<uint32_t>(edge->classification())));
+      feature.add_property(key_use,
+                           vtzero::encoded_property_value(static_cast<uint32_t>(edge->use())));
+      feature.add_property(key_tunnel, vtzero::encoded_property_value(edge->tunnel()));
+      feature.add_property(key_bridge, vtzero::encoded_property_value(edge->bridge()));
+      feature.add_property(key_roundabout, vtzero::encoded_property_value(edge->roundabout()));
+
+      // Check for opposing edge and add its metadata too (bidirectional roads)
+      baldr::graph_tile_ptr opp_tile = edge_tile;
```

Comment:
For bidirectional roads is it correct thing to do? Tbh I am not very well with these details of Valhalla...

Reply:
yep that's perfect